File: implement-shell-tools/ls/ls.py
#!/usr/bin/env python3
import sys
import os


def main():
    args = sys.argv[1:]
    show_all = False
    path = '.'

    for arg in args:
        if arg == '-1':
            pass  # one-per-line is our only output mode
        elif arg == '-a':
            show_all = True
        elif not arg.startswith('-'):
            path = arg

    try:
        entries = sorted(os.listdir(path), key=str.casefold)
    except OSError as e:
        print(f"ls: {path}: {e.strerror}", file=sys.stderr)
        sys.exit(1)

    if show_all:
        print('.')
        print('..')
        for entry in entries:
            print(entry)
    else:
        for entry in entries:
            if not entry.startswith('.'):
                print(entry)


if __name__ == '__main__':
    main()

File: implement-shell-tools/ls/ls.js
#!/usr/bin/env node
// Usage: node ls.js [-1] [-a] [path]
const fs = require('fs');

const args = process.argv.slice(2);
let showAll = false;
let targetPath = '.';

for (const arg of args) {
  if (arg === '-a') showAll = true;
  else if (arg === '-1') { /* one-per-line is our only output mode */ }
  else if (!arg.startsWith('-')) targetPath = arg;
}

const entries = fs.readdirSync(targetPath)
  .sort((a, b) => a.toLowerCase().localeCompare(b.toLowerCase()));

if (showAll) {
  console.log('.');
  console.log('..');
  for (const entry of entries) console.log(entry);
} else {
  for (const entry of entries) {
    if (!entry.startsWith('.')) console.log(entry);
  }
}
